cocotb/drivers/amba.py

```diff
@@ -220,98 +223,216 @@ class AXI4Slave(BusDriver):
         if AxSIZE < 7:
             return 2 ** AxSIZE
         return None
 
+    @cocotb.coroutine
+    def _saxi_handshake(self, clock, ready, valid, delay=0):
+        """ Wait for a valid handshake on the supplied AXI signals
+
+        delay specifies the clock cycles to wait before asserting ready
+        after the master asserts valid.
+            delay < 0: ready is asserted before valid.
+            delay = 0: ready asserted combinatorially with valid.
+            delay > 0: wait 'delay' clock cycles after valid to assert ready.
+        """
+        ready <= 1 if delay < 0 else 0
+
+        # check if valid is already asserted
+        if not valid.value:
+            yield RisingEdge(valid)
+
+        for i in range(delay):
+            yield RisingEdge(clock)
+
+        ready <= 1
+        yield RisingEdge(clock)
+        ready <= 0
+
+    @cocotb.coroutine
+    def _write_address(self):
+        """Monitor the Write Address channel for transactions
+        """
+        yield self._saxi_handshake(self.clock,
+                                   self.bus.AWREADY,
+                                   self.bus.AWVALID,
+                                   self.awready_delay)
+
+        awaddr = int(self.bus.AWADDR)
+        awlen = int(self.bus.AWLEN)
+        awsize = int(self.bus.AWSIZE)
+        awburst = int(self.bus.AWBURST)
+        awprot = int(self.bus.AWPROT)
+        awid = int(self.bus.AWID)
+
+        return awaddr, awlen, awsize, awburst, awprot, awid
+
     @cocotb.coroutine
     def _write_data(self):
-        clock_re = RisingEdge(self.clock)
+        """Monitor the Write Address channel for transactions
+        """
+        last = 0
+        wdata = []
+        wstrb = []
+        while not last:
+            yield self._saxi_handshake(self.clock,
+                                       self.bus.WREADY,
+                                       self.bus.WVALID,
+                                       self.wready_delay)
+
+            wdata.append(int(self.bus.WDATA))
+            wstrb.append(int(self.bus.WSTRB))
+            last = int(self.bus.WLAST)
+            wid = int(self.bus.WID)  # should be the same for each beat
+        return wdata, wid, wstrb
+
+    @cocotb.coroutine
+    def _write_transaction(self):
+        """Process write transactions
 
+        This requires both wdata and awaddr to complete.
+        TODO support out of order threads
+        """
         while True:
-            while True:
-                self.bus.WREADY <= 0
-                yield ReadOnly()
-                if self.bus.AWVALID.value:
-                    self.bus.WREADY <= 1
-                    break
-                yield clock_re
+            self.bus.BVALID <= 0
+            (
+                (wdata, wid, wstrb),
+                (awaddr, awlen, awsize, awburst, awprot, awid)
+            ) = yield Combine(
+                self._write_data(),
+                self._write_address()
+            )
```

Comment:
Apologies, i think I'm wrong - I don't think `Combine` works like this, nor did it in any previous version. I'm thinking of the trigger that I never finished writing. Here's the source for `Combine`:

https://github.com/potentialventures/cocotb/blob/ed3e6d2dd4e870c40462d4385cc27779979bc961/cocotb/triggers.py#L525-L549

Note that that just returns `self`, not the list of `ret.get()` values that could have been computed on line 544.